relengapi/blueprints/tooltool/grooming.py

```diff
@@ -0,0 +1,168 @@
+    subq = session.query(
+        fi_tbl.file_id,
+        sa.func.count('*').label('instance_count'))
+    subq = subq.group_by(fi_tbl.file_id)
+    subq = subq.subquery()
+    q = session.query(f_tbl)
+    q = q.join(subq, f_tbl.id == subq.c.file_id)
+    q = q.filter(subq.c.instance_count < num_regions)
+    q = q.all()
+    for file in q:
+        replicate_file(session, file)
+    session.commit()
+
+
+def replicate_file(session, file):
+    config = current_app.config['TOOLTOOL_REGIONS']
+    regions = set(config)
+    file_regions = set([i.region for i in file.instances])
+    # only use configured source regions; if a region is removed
+    # from the configuration, we can't copy from it.
+    source_regions = file_regions & regions
+    if not source_regions:
+        # this should only happen when the only region containing a
+        # file is removed from the configuration
+        log.warning("no source regions for {}".format(file.sha512))
+        return
+    source_region = source_regions.pop()
+    source_bucket = config[source_region]
+    target_regions = regions - file_regions
+    log.info("replicating {} from {} to {}".format(
+        file.sha512, source_region, ', '.join(target_regions)))
+
+    key_name = util.keyname(file.sha512)
+    for target_region in target_regions:
+        target_bucket = config[target_region]
+        conn = current_app.aws.connect_to('s3', target_region)
+        bucket = conn.get_bucket(target_bucket)
+        bucket.copy_key(new_key_name=key_name,
+                        src_key_name=key_name,
+                        src_bucket_name=source_bucket,
+                        storage_class='STANDARD',
+                        preserve_acl=False)
+        session.add(tables.FileInstance(file=file, region=target_region))
+
+
+@celery.task
+def check_file_pending_uploads(sha512):
+    """Check for pending uploads for a single file"""
+    session = current_app.db.session('tooltool')
+    file = tables.File.query.filter(tables.File.sha512 == sha512).first()
+    if file:
+        for pu in file.pending_uploads:
+            check_pending_upload(session, pu)
+    session.commit()
+
+
+def verify_file_instance(file, key):
+    """Verify that the given File table row and the given S3 Key match
+    in size and digest."""
+    if key.size != file.size:
+        log.warning("Uploaded file {} has unexpected size {}; expected "
+                    "{}".format(file.sha512, key.size, file.size))
+        return False
+
+    m = hashlib.sha512()
+    for bytes in key:
+        m.update(bytes)
+
+    if m.hexdigest() != file.sha512:
+        log.warning("Digest of file {} does not match".format(file.sha512))
+        return False
+
+    # verify some settings on the key, in case the uploader configured
+    # it differently
+    if key.storage_class != 'STANDARD':
+        return False
+
+    if key.get_redirect():  # pragma: no cover
+        # (moto doesn't support redirects)
+        return False
```

Comment:
Can you add log.warning() so we can see why it was False. Or maybe refactor to use exceptions instead.